src/XMakeBuildEngine/Evaluation/LazyItemEvaluator.cs

```diff
@@ -136,28 +137,68 @@ public ItemData(I item, int elementOrder, bool conditionResult)
         private class OperationWithCache : IItemOperation
         {
             private readonly IItemOperation _operation;
-            private readonly Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>> _cache;
+            private Dictionary<ISet<string>, ImmutableList<ItemData>> _cache;
+
+            private bool _isReferenced;
+#if DEBUG
+            private bool _applyCalled;
+#endif
 
             public OperationWithCache(IItemOperation operation)
             {
                 _operation = operation;
-                _cache = new Lazy<Dictionary<ISet<string>, ImmutableList<ItemData>>>(
-                    () => new Dictionary<ISet<string>, ImmutableList<ItemData>>());
             }
 
             public void Apply(ImmutableList<ItemData>.Builder listBuilder, ImmutableHashSet<string> globsToIgnore)
             {
+#if DEBUG
+                Debug.Assert(!_applyCalled, "Apply should only be called once. Otherwise caching is not working");
```

Comment:
It should be valid for `Apply` to be called multiple times if the `globsToIgnore` is different each time.